modules/dlsobackend/dlsobackend.cc

```diff
@@ -0,0 +1,722 @@
+{
+  DNSResourceRecord* rr = (DNSResourceRecord*)ptr;
+  rr->qtype = record->qtype;
+  string qname = string(record->qname, record->qname_len);
+  rr->qname = DNSName(qname);
+  rr->qclass = QClass::IN;
+  rr->content = string(record->content, record->content_len);
+  rr->ttl = record->ttl;
+  rr->auth = record->auth;
+  rr->scopeMask = record->scope_mask;
+  rr->domain_id = record->domain_id;
+}
+
+bool DlsoBackend::get(DNSResourceRecord& rr)
+{
+  bool success = api->get(api->handle, fill_cb, &rr);
+
+  if (!success) {
+    in_query = false;
+    return false;
+  }
+
+  return true;
+}
+
+void fill_meta_cb(const void* ptr, uint8_t value_len, const struct dns_value* values)
+{
+  std::vector<std::string>* meta = (std::vector<std::string>*)ptr;
+  for (uint8_t i = 0; i < value_len; i++) {
+    const struct dns_value* value = &values[i];
+    string value_s = string(value->value, value->value_len);
+
+    (*meta).push_back(value_s);
+  }
+}
+
+void fill_metas_cb(const void* ptr, uint8_t meta_len, const struct dns_meta* c_metas)
+{
+  std::map<std::string, std::vector<std::string>>* metas = (std::map<std::string, std::vector<std::string>>*)ptr;
+  for (uint8_t i = 0; i < meta_len; i++) {
+    const struct dns_meta* meta = &c_metas[i];
+    string property = string(meta->property, meta->property_len);
+    const struct dns_value* values = meta->values;
+
+    fill_meta_cb(&(*metas)[property], meta->value_len, values);
+  }
+}
+
+bool DlsoBackend::getAllDomainMetadata(const DNSName& name, std::map<std::string, std::vector<std::string>>& metas)
+{
+  if (d_dnssec == false)
+    return false;
+  if (api->get_metas == nullptr)
+    return false;
+
+  metas.clear();
+
+  string qname = name.toString();
+  return api->get_metas(api->handle, qname.size(), qname.c_str(), fill_metas_cb, &metas);
+}
+
+bool DlsoBackend::getDomainMetadata(const DNSName& name, const std::string& kind, std::vector<std::string>& meta)
+{
+  if (api->get_meta == nullptr)
+    return false;
+
+  meta.clear();
+
+  string qname = name.toString();
+  return api->get_meta(api->handle, qname.size(), qname.c_str(), kind.size(), kind.c_str(), fill_meta_cb, &meta);
+}
+
+bool DlsoBackend::setDomainMetadata(const DNSName& name, const std::string& kind, const std::vector<std::string>& meta)
+{
+  if (api->set_meta == nullptr)
+    return false;
+
+  struct dns_value* values = (struct dns_value*)calloc(meta.size(), sizeof(struct dns_value));
+  if (values == nullptr)
+    return false;
+  for (uint i = 0; i < meta.size(); i++) {
+    values[i].value_len = meta[i].size();
+    values[i].value = meta[i].c_str();
+  }
+
+  string qname = name.toString();
+  bool status = api->set_meta(api->handle, qname.size(), qname.c_str(), kind.size(), kind.c_str(), meta.size(), values);
+  free(values);
+  return status;
+}
+
+void fill_key_cb(const void* ptr, const struct dnskey* dnskey)
+{
+  std::vector<DNSBackend::KeyData>* keys = (std::vector<DNSBackend::KeyData>*)ptr;
```

Comment:
Also the const thing. the root of the problem is in the API `lib_so_api` that pass these opaque ptr as `const * void` it's an open question.

The intend is to clearly said to 3rd party lib implementer using this backend to avoid at all cost writing by themself to this pr.

Reply:
I would personally get rid of the `const` qualifier, as I find it misleading because the content is actually modified. Perhaps using opaque types would be better than a `void*`, but I have not reviewed the code to be sure it is possible to do so.

Reply:
The opaque type as in a C definition may be better, may explore it.

For the `const` qualifier get rid of it found it also confusing since we modify the data, may need some doc to express that this pointer need to remain untouched in all the callback-api.